storage/queue.go

```diff
@@ -111,6 +116,75 @@ type PeekMessageResponse struct {
 	MessageText    string `xml:"MessageText"`
 }
 
+// QueueMetadataResponse represents user defined metadata and queue
+// properties on a specific queue.
+//
+// See https://msdn.microsoft.com/en-us/library/azure/dd179384.aspx
+type QueueMetadataResponse struct {
+	ApproximateMessageCount int
+	UserDefinedMetadata     map[string]string
+}
+
+// SetMetadata operation sets user-defined metadata on the specified queue.
+// Metadata is associated with the queue as name-value pairs.
+//
+// See https://msdn.microsoft.com/en-us/library/azure/dd179384.aspx
+func (c QueueServiceClient) SetMetadata(name string, metadata map[string]string) error {
+	uri := c.client.getEndpoint(queueServiceName, pathForQueue(name), url.Values{"comp": []string{"metadata"}})
+	headers := c.client.getStandardHeaders()
+	headers["Content-Length"] = "0"
+	for k, v := range metadata {
+		headers[userDefinedMetadataHeaderPrefix+k] = v
+	}
+
+	resp, err := c.client.exec("PUT", uri, headers, nil)
+	if err != nil {
+		return err
+	}
+	defer resp.body.Close()
+
+	return checkRespCode(resp.statusCode, []int{http.StatusNoContent})
+}
+
+// GetMetadata operation retrieves user-defined metadata and queue
+// properties on the specified queue. Metadata is associated with
+// the queue as name-values pairs.
+//
+// See https://msdn.microsoft.com/en-us/library/azure/dd179384.aspx
+//
+// Because the way Golang's http client (and http.Header in particular)
+// canonicalize header names, the returned metadata names would always
+// be Titlecased (first letter in upper case, all the rest lower case)
+// no matter what case was used on the call to SetMetadata.
+func (c QueueServiceClient) GetMetadata(name string) (QueueMetadataResponse, error) {
+	qm := QueueMetadataResponse{}
+	qm.UserDefinedMetadata = make(map[string]string)
+	uri := c.client.getEndpoint(queueServiceName, pathForQueue(name), url.Values{"comp": []string{"metadata"}})
+	headers := c.client.getStandardHeaders()
+	resp, err := c.client.exec("GET", uri, headers, nil)
+	if err != nil {
+		return qm, err
+	}
+	defer resp.body.Close()
+
+	for k, v := range resp.headers {
+		if len(v) != 1 {
+			return qm, fmt.Errorf("Unexpected number of values (%d) in response header '%s'", len(v), k)
+		}
+
+		if k == approximateMessagesCountHeader {
+			qm.ApproximateMessageCount, err = strconv.Atoi(v[0])
+			if err != nil {
+				return qm, fmt.Errorf("Unexpected value in response header '%s': '%s' ", k, v[0])
+			}
+		} else if strings.Index(k, userDefinedMetadataHeaderPrefix) == 0 {
+			qm.UserDefinedMetadata[k[len(userDefinedMetadataHeaderPrefix):]] = v[0]
```

Comment:
now this makes me think if we should just return `http.Header`s  with the user defined metadata keys and use its `.Get` method etc... This notice is actually fine, and people should be testing their code, but this will be hard to change in the future if need be.

As far as I observed so far, our SDKs return lowercase names (as documented in the REST API Docs) for the headers... Therefore: **I think we should lowercase all of them before putting to the map.** This would be less surprising for the user and they'd notice the pattern on the first use and we can sustain this behavior more easily.

Title-case is just very confusing. English is my second language and for instance I can't say I fully get what you mean by the following:

> Title-cased (first letter in upper case, all the rest lower case)

to me, that sounds like `X-ms-meta-foo`; not `X-Ms-Meta-Foo`. I know it's obvious what title-case is but my preference would be making it all lowercase.